src/audio/crossover/crossover.c

```diff
@@ -0,0 +1,745 @@
+		if (!cd->config_new) {
+			comp_err(dev, "crossover_cmd_set_data(), alloc fail");
+			return -EINVAL;
+		}
+
+		/* Copy the configuration. If the component state is ready
+		 * the Crossover will initialize in prepare().
+		 */
+		ret = memcpy_s(cd->config_new, bs, cdata->data->data, bs);
+		assert(!ret);
+
+		/* If component state is READY we can omit old configuration
+		 * immediately. When in playback/capture the new configuration
+		 * presence is checked in copy().
+		 */
+		if (dev->state == COMP_STATE_READY)
+			crossover_free_config(&cd->config);
+
+		/* If there is no existing configuration the received can
+		 * be set to current immediately. It will be applied in
+		 * prepare() when streaming starts.
+		 */
+		if (!cd->config) {
+			cd->config = cd->config_new;
+			cd->config_new = NULL;
+		}
+
+		break;
+	default:
+		comp_err(dev, "crossover_cmd_set_data(), invalid command");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/**
+ * \brief Handles incoming IPC commands for Crossover component.
+ */
+static int crossover_cmd(struct comp_dev *dev, int cmd, void *data,
+			 int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+		comp_info(dev, "crossover_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = crossover_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = crossover_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	default:
+		comp_err(dev, "crossover_cmd(), invalid command");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+/**
+ * \brief Sets Crossover Filter component state.
+ * \param[in,out] dev Crossover Filter base component device.
+ * \param[in] cmd Command type.
+ * \return Error code.
+ */
+static int crossover_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "crossover_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+/**
+ * \brief Copies and processes stream data.
+ * \param[in,out] dev Crossover Filter base component device.
+ * \return Error code.
+ */
+static int crossover_copy(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	struct comp_buffer *sinks[CROSSOVER_MAX_STREAMS] = { NULL };
+	int i;
+	uint32_t num_sinks = 0;
+	uint32_t frames = -1;
+	uint32_t source_bytes;
+	uint32_t avail;
+	uint32_t flags = 0;
+	uint32_t sinks_bytes[CROSSOVER_MAX_STREAMS] = { 0 };
+
+	comp_dbg(dev, "crossover_copy()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 sink_list);
+
+	/* Check for changed configuration */
+	if (cd->config_new) {
+		crossover_free_config(&cd->config);
+		cd->config = cd->config_new;
+		cd->config_new = NULL;
+		/* TODO update coefficients with new config */
+	}
+
+	/* Use the assign_sink array from the config to route
+	 * the output to the corresponding sinks
+	 */
+	num_sinks = crossover_assign_sinks(dev, cd->config, sinks);
+
+	buffer_lock(source, &flags);
+
+	/* check if source is active */
+	if (source->source->state != dev->state) {
```

Comment:
How far does this go? Could we potentially be the first element and hit a null dereference? This seems a bit weird.

Reply:
It would be an issue if we can have pipelines with no PCM or DAI? 
normally we have:
PCM -> B0 -> Crossover -> B1 -> DAI.

Could we potentially have this?:
Crossover -> B0 -> DAI

Maybe that could be a possibility if we were to "concatenate" pipelines together. An example would be sof-cml-demux-rt5682.m4.

Reply:
sounds good to me, we will see how upstream likes it